oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/R2dbcReactiveOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,370 @@
+
+	private static final String PK_FILTER = "client_registration_id = :clientRegistrationId AND principal_name = :principalName";
+
+	// @formatter:off
+	private static final String LOAD_AUTHORIZED_CLIENT_SQL = "SELECT " + COLUMN_NAMES + " FROM " + TABLE_NAME
+			+ " WHERE " + PK_FILTER;
+	// @formatter:on
+
+	// @formatter:off
+	private static final String SAVE_AUTHORIZED_CLIENT_SQL = "INSERT INTO " + TABLE_NAME + " (" + COLUMN_NAMES + ")" +
+			"VALUES (:clientRegistrationId, :principalName, :accessTokenType, :accessTokenValue," +
+				" :accessTokenIssuedAt, :accessTokenExpiresAt, :accessTokenScopes, :refreshTokenValue," +
+				" :refreshTokenIssuedAt)";
+	// @formatter:on
+
+	private static final String REMOVE_AUTHORIZED_CLIENT_SQL = "DELETE FROM " + TABLE_NAME + " WHERE " + PK_FILTER;
+
+	// @formatter:off
+	private static final String UPDATE_AUTHORIZED_CLIENT_SQL = "UPDATE " + TABLE_NAME +
+			" SET access_token_type = :accessTokenType, " +
+			" access_token_value = :accessTokenValue, " +
+			" access_token_issued_at = :accessTokenIssuedAt," +
+			" access_token_expires_at = :accessTokenExpiresAt, " +
+			" access_token_scopes = :accessTokenScopes," +
+			" refresh_token_value = :refreshTokenValue, " +
+			" refresh_token_issued_at = :refreshTokenIssuedAt" +
+			" WHERE " +
+			PK_FILTER;
+	// @formatter:on
+
+	protected final DatabaseClient databaseClient;
+
+	protected Function<OAuth2AuthorizedClientHolder, Map<String, Parameter>> authorizedClientParametersMapper;
+
+	protected BiFunction<Row, RowMetadata, Mono<OAuth2AuthorizedClient>> authorizedClientRowMapper;
+
+	/**
+	 * Constructs a {@code R2dbcReactiveOAuth2AuthorizedClientService} using the provided
+	 * parameters.
+	 * @param databaseClient the DatabaseClient
+	 * @param clientRegistrationRepository the repository of client registrations
+	 */
+	public R2dbcReactiveOAuth2AuthorizedClientService(DatabaseClient databaseClient,
+			ReactiveClientRegistrationRepository clientRegistrationRepository) {
+		Assert.notNull(databaseClient, "databaseClient cannot be null");
+		Assert.notNull(clientRegistrationRepository, "clientRegistrationRepository cannot be null");
+		this.databaseClient = databaseClient;
+		this.authorizedClientParametersMapper = new OAuth2AuthorizedClientParametersMapper();
+		this.authorizedClientRowMapper = new OAuth2AuthorizedClientRowMapper(clientRegistrationRepository);
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T extends OAuth2AuthorizedClient> Mono<T> loadAuthorizedClient(String clientRegistrationId,
+			String principalName) {
+		Assert.hasText(clientRegistrationId, "clientRegistrationId cannot be empty");
+		Assert.hasText(principalName, "principalName cannot be empty");
+
+		return (Mono<T>) this.databaseClient.sql(LOAD_AUTHORIZED_CLIENT_SQL)
+				.bind("clientRegistrationId", clientRegistrationId).bind("principalName", principalName)
+				.map(this.authorizedClientRowMapper::apply).first().flatMap((authorizedClient) -> authorizedClient);
```

Comment:
Agreed, I'm also not in favour of the `flatMap()` that is required because of the `Mono<OAuth2AuthorizedClient>` return type in the row mapper.

Let's revert the return type to `OAuth2AuthorizedClientHolder` having these members:

```java
	public static final class OAuth2AuthorizedClientHolder {
		private final String clientRegistrationId;
		private final String principalName;
		private final OAuth2AccessToken accessToken;
		private final OAuth2RefreshToken refreshToken;


```

Reply:
This means that the ```authorizedClientParametersMapper``` will also be reverted to 
`
private BiFunction<OAuth2AuthorizedClient, Authentication, Map<String, Parameter>> authorizedClientParametersMapper;
`
or do you prefer using the same signature

Reply:
```
protected Function<OAuth2AuthorizedClientHolder, Map<String, Parameter>> authorizedClientParametersMapper;

protected BiFunction<Row, RowMetadata, OAuth2AuthorizedClientHolder> authorizedClientRowMapper;
```

and two constructors in ```OAuth2AuthorizedClientHolder```
```
public OAuth2AuthorizedClientHolder(OAuth2AuthorizedClient authorizedClient, Authentication principal) {
			Assert.notNull(authorizedClient, "authorizedClient cannot be null");
			Assert.notNull(principal, "principal cannot be null");
			this.clientRegistrationId = authorizedClient.getClientRegistration().getRegistrationId();
			this.principalName = principal.getName();
			this.accessToken = authorizedClient.getAccessToken();
			this.refreshToken = authorizedClient.getRefreshToken();
		}

		public OAuth2AuthorizedClientHolder(String clientRegistrationId, String principalName,
				OAuth2AccessToken accessToken, OAuth2RefreshToken refreshToken) {
			this.clientRegistrationId = clientRegistrationId;
			this.principalName = principalName;
			this.accessToken = accessToken;
			this.refreshToken = refreshToken;
		}
```
The first one is used when saving an authorized client. The second one is used when loading an authorized client.

Reply:
Let's use `OAuth2AuthorizedClientHolder` in both `authorizedClientParametersMapper` and `authorizedClientRowMapper`. Using a wrapper object in the signature provides us the flexibility in "adding" members to the wrapper object (if later required) without breaking the `BiFunction` signature.

The two constructors in `OAuth2AuthorizedClientHolder` makes sense. I would delegate the `OAuth2AuthorizedClient` constructor to the other constructor.

Reply:
I just noticed that there are a lot more changed files in latest commit. Can you ensure that you **rebase** on top of master and not merge.

Reply:
@jgrandja I did a rebase. But, because there was a breaking API change in spring-framework, test module, I had to update Kotlin tests from spring-security-config (https://github.com/spring-projects/spring-framework/commit/d04c5f8b2cb274b8f44c47b3783e5d29f3e21b43#diff-5a70a1d6caaa66c8feec94bbaf723ee682cdafda132c988ef46a365cd6886f62)

Reply:
> The two constructors in `OAuth2AuthorizedClientHolder` makes sense. I would delegate the `OAuth2AuthorizedClient` constructor to the other constructor.

By doing this, following checks
```
Assert.notNull(authorizedClient, "authorizedClient cannot be null");
Assert.notNull(principal, "principal cannot be null");
```
will no longer be in place since the delegate constructor call must be the first statement in a constructor

Reply:
Looks like it has been fixed, see https://github.com/spring-projects/spring-security/commit/9cf3129d7afa2abb439aba6aadfee0a2c8c784bf

Reply:
I see. Ok then don't bother with delegating constructor as the `Assert` is necessary

Reply:
> Looks like it has been fixed, see [9cf3129](https://github.com/spring-projects/spring-security/commit/9cf3129d7afa2abb439aba6aadfee0a2c8c784bf)

yep, right. Looks like I rebased before it was fixed in the main repository :) I will revert the changes